### PR TITLE
Make attached particle spawners respect the parent's yaw

### PR DESCRIPTION
**Pull request by orwell96**
The position values of attached particle spawners are added to the parent's position absolutely. This is no problem for particles spawned equally around the parent's position, but as soon as you want to spawn particles in front or behind of objects, you run into problems.
Example: you want smoke coming out of a steam engine, but every time the engine's yaw changes you need to reinitialize the particle spawner.
With this PR, position, velocity and acceleration vectors are rotated by the yaw of the parent object so that they are truly relative to it.

### CODE REVIEW

File: src/clientobject.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ class ClientActiveObject : public ActiveObject
 	virtual bool getCollisionBox(aabb3f *toset){return false;}
 	virtual bool collideWithObjects(){return false;}
 	virtual v3f getPosition(){return v3f(0,0,0);}
+	virtual float getYaw(){return 0;}
```
Comment (nerzhul) [inline src/clientobject.h]: this function should be const

File: src/content_cao.cpp [hunk 2]
```diff
@@ -690,6 +691,10 @@ v3f GenericCAO::getPosition()
 	}
 	return pos_translator.vect_show;
 }
+float GenericCAO::getYaw()
```
Comment (nerzhul) [inline src/content_cao.cpp]: please inline this function

File: src/clientobject.h (hunk 1 above)
Comment (sfan5) [inline src/clientobject.h]: all others aren't const either, but i agree that these can be const

Comment (orwell96) [inline src/clientobject.h]: so either make both const, or do nothing, right?
should I make them const?

**Comment by raymoo** [comment]
Would it be possible to also rotate velocity vectors? For things like breath effects.

**Comment by paramat** [comment]
Is this acceptable seeing that the pos, vel, acc values for the particles are not being rotated?

**Comment by raymoo** [comment]
Actually I was mistaken, and didn't even completely read the PR. Velocity values are already rotated.

**Comment by orwell96** [comment]
@paramat: They are, this is entirely what it does.

**Comment by paramat** [comment]
Good.

File: src/content_cao.cpp (hunk 2 above)
Comment (nerzhul) [inline src/content_cao.cpp]: by saying making inline here i would say define the function with declaration

Comment (orwell96) [inline src/content_cao.cpp]: Sorry, I am confused. Function is declaed and defined. Do you mean
inline float GenericCAO::getYaw() {return m_yaw;}
or something different?

**Comment by orwell96** [comment]
https://github.com/minetest/minetest/commit/3db2f08ff95a2e1366b9b5079100b6a681f1e562
may introduce merge conflicts!

**Comment by nerzhul** [comment]
@orwell96 just pull rebase to fix the merge problem. github doesn't say there is problem here

File: src/content_cao.cpp (hunk 2 above)
Comment (nerzhul) [inline src/content_cao.cpp]: exact (don't forget spaces)

**Comment by orwell96** [comment]
I have tested it, it works without problems.
screenshot_20161026_152912
Wait with merge, I will sync the fork and merge recent change first.

**Comment by orwell96** [comment]
OK, no conflicts, ready to merge.

File: src/content_cao.cpp (hunk 2 above)
Comment (orwell96) [inline src/content_cao.cpp]: and why isn't this done with getPosition()?

Comment (paramat) [inline src/content_cao.cpp]: Perhaps because that function is longer.

**Comment by paramat** [comment]
Perhaps add a line to lua_api.txt stating yaw behaviour?

**Comment by orwell96** [comment]
OK.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -4121,7 +4121,7 @@ The Biome API is still in an experimental phase and subject to change.
     --  ^ collision_removal: if true then particle is removed when it collides,
     --  ^ requires collisiondetection = true to have any effect
         attached = ObjectRef,
-    --  ^ attached: if defined, makes particle positions relative to this object.
+    --  ^ attached: if defined, Particle positions, velocities and accelerations are relative to this object's position and yaw.
```
Comment (paramat) [inline doc/lua_api.txt]: 'Particle' doesn't need a capital letter.
Please split this long line at <=80 columns.

**Comment by orwell96** [comment]
about the const above: where to put it, I'm not sure if I would break it.

**Comment by paramat** [comment]
I think best to not change anything to const unless you get more review that states exactly what to make const.
@nerzhul @sfan5 could you specify what functions can be const?

**Comment by orwell96** [comment]
getYaw() is now const.
I didn't touch the other functions.

**Comment by nerzhul** [comment]
The getYaw function should be const

File: src/content_cao.cpp [hunk 4]
```diff
@@ -690,6 +691,10 @@ v3f GenericCAO::getPosition()
 	}
 	return pos_translator.vect_show;
 }
+inline float GenericCAO::getYaw() const
```
Comment (nerzhul) [inline src/content_cao.cpp]: move this definition to class instead, it's only 1 line

**Comment by orwell96** [comment]
Sorry if I get annoying, but I think this should be merged soon, before mod developers implement the particles with the current behavior
(this is the case for some other features which were not finally done and then never were because of 'compatibility')

**Comment by paramat** [comment]
I assume @nerzhul approves the PR, added label.

**Comment by paramat** [comment]
Looking at casually it it seems good codewise.

**Comment by orwell96** [comment]
No, please squash during merge. It's to much messing around with history for me.

File: src/clientobject.h [hunk 5]
```diff
@@ -61,6 +61,9 @@ class ClientActiveObject : public ActiveObject
 	virtual bool getCollisionBox(aabb3f *toset){return false;}
 	virtual bool collideWithObjects(){return false;}
 	virtual v3f getPosition(){return v3f(0,0,0);}
+	virtual float getYaw() const {
+		return 0;
+	}
```
Comment (sfan5) [inline src/clientobject.h]: should be a single line

File: src/particles.cpp [hunk 6]
```diff
@@ -253,11 +253,15 @@ void ParticleSpawner::step(float dtime, ClientEnvironment* env)
...
 	if (m_attached_id != 0) {
-		if (ClientActiveObject *attached = env->getActiveObject(m_attached_id))
-			attached_offset = attached->getPosition() / BS;
-		else
+		if (ClientActiveObject *attached = env->getActiveObject(m_attached_id)) {
+			attached_pos = attached->getPosition() / BS;
+			attached_yaw = attached->getYaw();
+			is_attached = true;
+		} else
 			unloaded = true;
```
Comment (sfan5) [inline src/particles.cpp]: braces should be used consistently in a single if-statement (add them here)

**Comment by paramat** [comment]
Applying: Make attached particle spawners respect the parent's yaw when spawning particles
error: patch failed: src/particles.cpp:274
error: src/particles.cpp: patch does not apply

**Comment by orwell96** [comment]
Seems line I can't reopen it.

**Comment by raymoo** [comment]
Is it because you deleted the branch?

**Comment by orwell96** [comment]
#4782

**Pull request closed without merging**